File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Loader/SpineCharacterDisplay.ts
import { Spine, FairyGUI, UnityEngine } from "csharp";
import { Rect } from "../../../../Core/Define/Rect";
import Color from "../DataDefine/Color";
import { View_UnitWrap } from "../Gen/View/ModTheSpire_Combat";
import { ViewComponent_HealthBar } from "../UI/ViewComponent/ViewComponent_HealthBar";
import { IDisplay } from "./IDisplay";

export class SpineCharacterDisplay implements IDisplay{
    get FlipX(): boolean {
        return this.Self.Skeleton.ScaleX == -1;
    }

    set FlipX(value: boolean) {
        value ? this.Self.Skeleton.ScaleX = -1 : 1;
    }

    get FlipY(): boolean {
        return this.Self.Skeleton.ScaleY == -1;
    }

    set FlipY(value: boolean) {
        value ? this.Self.Skeleton.ScaleY = -1 : 1;
    }
    get Visible(): boolean {
        return this._Visible;
    }

    set Visible(value: boolean) {
        this._Visible = value;
        this._CacheGameObject.SetActive(value);
    }
    get Rotation(): number {
        return this._Rotation;
    }

    set Rotation(value: number) {
        this._Rotation = value;
        this._Wrap.rotation = value;
    }

    get ScaleX(): number {
        return this._ScaleX;
    }

    set ScaleX(value: number) {
        this._ScaleX = value;
        this._Wrap.scaleX = value;
    }

    get ScaleY(): number {
        return this._ScaleY;
    }

    set ScaleY(value: number) {
        this._ScaleY = value;
        this._Wrap.scaleY = value;
    }

    get Self(): Spine.Unity.SkeletonAnimation {
        return this._Self;
    }

    get SortingOrder(): number {
        return this._SortingOrder;
    }

    set SortingOrder(value: number) {
        this._SortingOrder = value;
        this.UnitComponent.sortingOrder = value;
    }

    get X(): number {
        return this._X;
    }

    set X(value: number) {
        this._X = value;
        this._Wrap.x = value;
    }

    get Y(): number {
        return this._Y;
    }

    set Y(value: number) {
        this._Y = value;
        this._Wrap.y = value;
    }

    get Color(): Color {
        return this._Color;
    }

    set Color(value: Color) {
        this._Color = value;
        Spine.Unity.SkeletonExtensions.SetColor(this.Self.Skeleton,value.UnityColor);
    }
    
    public Bounds: FairyGUI.GGraph;
    private _Rotation: number;
    private _ScaleX: number;
    private _ScaleY: number;
    private _SortingOrder: number;
    private _X: number;
    private _Y: number;
    private _Visible: boolean;
    private _Color: Color;
    private _FlipX: boolean;
    private _FlipY: boolean;
    private readonly _Self: Spine.Unity.SkeletonAnimation;
    UnitComponent:FairyGUI.GComponent;
    HealthComponent: ViewComponent_HealthBar;
    private readonly _Wrap : FairyGUI.GoWrapper;
    private readonly _CacheGameObject:UnityEngine.GameObject;
    private readonly _CacheTransform:UnityEngine.Transform;
    constructor(animation : Spine.Unity.SkeletonAnimation) {
        this._Self = animation; 
        this._CacheGameObject = animation.gameObject;
        this._CacheTransform = animation.transform;
        this.UnitComponent = <FairyGUI.GComponent>View_UnitWrap.CreateInstance();
        this._Wrap = new FairyGUI.GoWrapper();
        this._Wrap.SetWrapTarget(animation.gameObject,true);
        this.UnitComponent.GetChild("ModelLoader").asGraph.SetNativeObject(this._Wrap);
        this.HealthComponent = new ViewComponent_HealthBar(this.UnitComponent.GetChild("HealthBar").asProgress);
        this.Bounds = this.UnitComponent.GetChild("Bounds").asGraph;
        this.Bounds.SetPivot(0.5,0.5,true);
        this.Rotation = 0;
        this.ScaleX = 1;
        this.ScaleY = 1;
        this.SortingOrder = 0;
        this.X = 0;
        this.Y = 0;
        this.Visible = true;
        this.Color = Color.White;
    }

    AddAnimation(animation: string, loop: boolean = false,delay:number = 0) {
        this._Self.AnimationState.AddAnimation(0,animation,loop,delay);
    }

    SetAnimation(animation: string, loop:boolean = false) : any {
        this._Self.AnimationName = animation;
        return this._Self.AnimationState.SetAnimation(0,animation,loop);
    }

    Dispose() {
        this.UnitComponent.Dispose();
        this._Wrap.Dispose();
        UnityEngine.Object.Destroy(this._Self.gameObject);
    }

    SetParent(object: FairyGUI.GComponent) {
        if(object){
            object.AddChild(this.UnitComponent);
        }
    }

    PlayFastAttack() {
        if(!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing)) {
            let transition = this.UnitComponent.GetTransition("FastAttack");
            if(!transition.playing)
                transition.Play();
        }
    }

    PlayHop() {
        if(!(this.UnitComponent.GetTransition("Jump").playing)){
            let transition = this.UnitComponent.GetTransition("Hop");
            if(!transition.playing)
                transition.Play();
        }
    }

    PlayJump() {
        if(!(this.UnitComponent.GetTransition("Hop").playing)){
            let transition = this.UnitComponent.GetTransition("Jump");
            if(!transition.playing)
                transition.Play();
        }
    }

    PlaySlowAttack() {
        if(!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing)){
            let transition = this.UnitComponent.GetTransition("SlowAttack");
            if(!transition.playing)
                transition.Play();
        }
    }

    PlaySlowShake(duration:number) {
        if(!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing)){
            let transition = this.UnitComponent.GetTransition("SlowShake");
            if(!transition.playing) {
                transition.Play(duration / 0.1,0,null);
            }
        }
    }

    PlayFastShake(duration:number) {
        if(!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing)){
            let transition = this.UnitComponent.GetTransition("FastShake");
            if(!transition.playing)
                transition.Play(duration / 0.1,0,null);
        }
    }

    PlayStagger() {
        if(!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing)){
            let transition = this.UnitComponent.GetTransition("Stagger");
            if(!transition.playing)
                transition.Play();
        }
    }

    SetHitBox(rect:Rect) {
        if(rect.X)
            this.Bounds.x = rect.X;
        if(rect.Y)
            this.Bounds.y = rect.Y;
        if(rect.Width)
            this.Bounds.width = rect.Width;
        if(rect.Height)
            this.Bounds.height = rect.Height;
    }

    GetHitBox():Rect{
        return new Rect(this.Bounds.x,this.Bounds.y,this.Bounds.width,this.Bounds.height);
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Unit/Loader/SpineCharacterDisplay.cs
using FairyGUI;
using ModTheSpire.ModTheSpire_Combat;
using ModTheSpire.UI.ViewComponent;
using Spine.Unity;
using UnityEngine;

namespace ModTheSpire.Unit.Loader
{
    public class SpineCharacterDisplay : IDisplay
    {
        private SkeletonAnimation SkeletonAnimation;
        private UnityObject Uo;
        private GoWrapper mWrapper;

        public float Alpha
        {
            get => SkeletonAnimation.Skeleton.A;
            set => SkeletonAnimation.Skeleton.A = value;
        }

        public float X
        {
            get => Uo.X;
            set => Uo.X = value;
        }

        public float Y
        {
            get => Uo.Y;
            set => Uo.Y = value;
        }

        public Vector3 Rotation
        {
            get => Uo.EulerAngles;
            set => Uo.EulerAngles = value;
        }

        public float ScaleX
        {
            get => Uo.LocalScale.x;
            set => Uo.LocalScaleX = value;
        }

        public float ScaleY
        {
            get => Uo.LocalScale.y;
            set => Uo.LocalScaleY = value;
        }

        public int SortingOrder
        {
            get => this.UnitComponent.sortingOrder;
            set => this.UnitComponent.sortingOrder = value;
        }

        public bool Visible
        {
            get => this.UnitComponent.visible;
            set => this.UnitComponent.visible = value;
        }

        public Color32 Color
        {
            get => this.SkeletonAnimation.Skeleton.GetColor();
            set => this.SkeletonAnimation.Skeleton.SetColor(value);
        }

        public bool FlipX
        {
            get => SkeletonAnimation.Skeleton.ScaleX < 0;
            set => SkeletonAnimation.Skeleton.ScaleX *= -1;
        }

        public bool FlipY
        {
            get => SkeletonAnimation.Skeleton.ScaleY < 0;
            set => SkeletonAnimation.Skeleton.ScaleY *= -1;
        }

        public GComponent UnitComponent { get; set; }
        public GGraph Bounds { get; set; }
        public ViewComponent_HealthBar HealthComponent { get; set; }
        public object Self { get; }
        public GObject PowerComponent { get; set; }
        public GObject NameComponent { get; set; }

        public SpineCharacterDisplay(SkeletonAnimation animation)
        {
            Self = SkeletonAnimation = animation;
            Uo = new UnityObject(SkeletonAnimation.gameObject);
            UnitComponent = View_UnitWrap.CreateInstance();
            mWrapper = new GoWrapper();
            mWrapper.SetWrapTarget(animation.gameObject, true);
            UnitComponent.GetChild("ModelLoader").asGraph.SetNativeObject(this.mWrapper);
            HealthComponent = new ViewComponent_HealthBar(this.UnitComponent.GetChild("HealthBar").asProgress);
            PowerComponent = this.UnitComponent.GetChild("PowerList");
            NameComponent = this.UnitComponent.GetChild("DisplayName");
            Bounds = this.UnitComponent.GetChild("Bounds").asGraph;
            Bounds.SetPivot(0.5f, 0.5f, true);
            Rotation = Vector3.zero;
            ScaleX = 1;
            ScaleY = 1;
            SortingOrder = 0;
            X = 0;
            Y = 0;
            Visible = true;
            Color = UnityEngine.Color.white;
        }

        public void PlayFastAttack()
        {
            if (!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing))
            {
                var transition = this.UnitComponent.GetTransition("FastAttack");
                if (!transition.playing)
                    transition.Play();
            }
        }

        public void PlayHop()
        {
            if (!(this.UnitComponent.GetTransition("Jump").playing))
            {
                var transition = this.UnitComponent.GetTransition("Hop");
                if (!transition.playing)
                    transition.Play();
            }
        }

        public void PlayJump()
        {
            if (!(this.UnitComponent.GetTransition("Hop").playing))
            {
                var transition = this.UnitComponent.GetTransition("Jump");
                if (!transition.playing)
                    transition.Play();
            }
        }

        public void PlaySlowAttack()
        {
            if (!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing))
            {
                var transition = this.UnitComponent.GetTransition("SlowAttack");
                if (!transition.playing)
                    transition.Play();
            }
        }

        public void PlaySlowShake(float duration)
        {
            if (!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing))
            {
                var transition = this.UnitComponent.GetTransition("SlowShake");
                if (!transition.playing)
                {
                    transition.Play();
                }
            }
        }

        public void PlayFastShake(float duration)
        {
            if (!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing))
            {
                var transition = this.UnitComponent.GetTransition("FastShake");
                if (!transition.playing)
                    transition.Play();
            }
        }

        public void PlayStagger()
        {
            if (!(this.UnitComponent.GetTransition("Jump").playing || this.UnitComponent.GetTransition("Hop").playing))
            {
                var transition = this.UnitComponent.GetTransition("Stagger");
                if (!transition.playing)
                    transition.Play();
            }
        }

        public void SetHitBox(Rect rect)
        {
             this.Bounds.x = rect.x;
             this.Bounds.y = rect.y;
             this.Bounds.width = rect.width;
             this.Bounds.height = rect.height;
        }

        public Rect GetHitBox() {
            return new Rect(this.Bounds.x, this.Bounds.y, this.Bounds.width, this.Bounds.height);
        }

        public void Dispose()
        {
            this.UnitComponent.Dispose();
            this.mWrapper.Dispose();
            Uo.Dispose();
        }
        
        public void SetParent(GComponent p)
        {
            if (p != null)
            {
                p.AddChild(this.UnitComponent);
            }
        }

        public object SetAnimation(string animation, bool loop = false)
        {
            return SkeletonAnimation.AnimationState.SetAnimation(0, animation, loop);
        }

        public object AddAnimation(string animation, bool loop = false, float delay = 0)
        {
            return SkeletonAnimation.AnimationState.AddAnimation(0, animation, loop, delay);
        }
    }
}